scarlet/blend.py

```diff
@@ -0,0 +1,659 @@
+                else:
+                    w = self._weights[bb_m].flatten()[:,None]
+                    ddx,ddy = np.dot(np.dot(np.linalg.inv(np.dot(MT, MT.T*w)), MT), y[bb_m].flatten())
+                if ddx**2 + ddy**2 > self.center_min_dist**2:
+                    center = source.center + (ddy, ddx)
+                    source.set_center(center)
+                    msg = "shifting source {0} by ({1}/2}) to (3}/4})"
+                    logger.info(msg.format(m, ddy, ddx, source.center[0], source.center[1]))
+
+    def _get_shift_differential(self, m):
+        """Calculate the difference image used ot fit positions
+
+        Parameters
+        ----------
+        m: int
+            Index of the source in `Blend.sources`
+
+        Returns
+        -------
+        diff_img: `~numpy.array`
+            Difference image in each band used to fit the position
+        """
+        # compute (model - dxy*shifted_model)/dxy for first-order derivative
+        source = self.sources[m]
+        slice_m = source.get_slice_for(self._img.shape)
+        k = self.component_of(m, 0)
+        model_m = self._models[k][self.sources[m].bb].copy()
+        # in self._models, per-source components aren't combined,
+        # need to combine here
+        for k in range(1,source.K):
+            model_m += self._models[k][self.sources[m].bb]
+
+        # get Gamma matrices of source m with additional shift
+        offset = source.shift_center
+        dx = source.center - source.center_int
+        pos_x = dx + (0, offset)
+        pos_y = dx + (offset, 0)
+        dGamma_x = source._gammaOp(pos_x, source.shape)
+        dGamma_y = source._gammaOp(pos_y, source.shape)
+        diff_img = [source.get_model(combine=True, Gamma=dGamma_x),
+                    source.get_model(combine=True, Gamma=dGamma_y)]
+        diff_img[0] = (model_m-diff_img[0][slice_m])/source.shift_center
+        diff_img[1] = (model_m-diff_img[1][slice_m])/source.shift_center
+        return diff_img
+
+    def _set_edge_flux(self, m, model):
+        """Keep track of the flux at the edge of the model
+
+        Parameters
+        ----------
+        m: int
+            Index of the source
+        model: `~numpy.array`
+            (Band,Height,Width) array of the model.
+
+        Returns
+        -------
+        None, but sets `self._edge_flux`
+        """
+        try:
+            self._edge_flux
+        except AttributeError:
+            self._edge_flux = np.zeros((self.M, 4, self.B))
+
+        # top, right, bottom, left
+        self._edge_flux[m,0,:] = np.abs(model[:,:,-1,:]).sum(axis=0).mean(axis=1)
+        self._edge_flux[m,1,:] = np.abs(model[:,:,:,-1]).sum(axis=0).mean(axis=1)
+        self._edge_flux[m,2,:] = np.abs(model[:,:,0,:]).sum(axis=0).mean(axis=1)
+        self._edge_flux[m,3,:] = np.abs(model[:,:,:,0]).sum(axis=0).mean(axis=1)
+
+    def resize_sources(self):
+        """Resize frames for sources (if necessary)
```

Comment:
We might want to come up with a better way to estimate how much larger the source needs to grow. That would help in cases where you have a very bright object that needs a much larger footprint.

Reply:
I agree. I tried to do this with the relative 25% increase. It's not trivial to know how to do the increase because we have no immediate way of knowing how much larger a source could be.